### DISCUSSION TITLE
How can i implement an Auth 2.0 authentication method?

### DISCUSSION THREAD

**Discussion by waapatricio** [saloonphp/saloon#61] [Q&A] [answered]
Reading the documentation I didn't find any examples. And it became a little confusing for me.

**Comment by Sammyjo20** [comment]
I am currently working on standardising the Oauth 2 process but for now you will need to make separate requests for each token request.  

You can use Saloon Authenticators to carry the access token, refresh token and expiry around with you and then I would recommend creating a plugin which checks if the authentication has expired or not. 

I really want to try and make Oauth 2 an easy process for Saloon, so I’m totally open to suggestions.

**Reply by juse-less** [reply]
I think the best bet overall is to use [The PHP League's oAuth Client package](https://github.com/thephpleague/oauth2-client).
AFAIK, most people use their Server package (including Laravel Passport).
Continuing on that, Laravel Socialite does have The PHP League's oAuth 1.0 package as a dependency, but unsure on how deeply they use it.
oAuth 2 is completely custom, however. It could be worth checking out their codebase for ideas and inspiration.
Both Laravel Socialite and The PHP League's oAuth 1.0 and oAuth 2.0 packages support a Provider approach, so you can easily define your service(s).

Since you've already got the `AuthenticatorInterface`, I'd probably create one for the Authorization Code Grant, and Client Credential Grant.
Looking at the Basic-, Digest-, and Token Authenticators, respectively, you're already requesting the user to insert the relevant data/credentials in the constructors.
So, from that standpoint, you'd easily hand off those responsibilities to the user to input the Access Token/Refresh Tokens from the end users browser, or the client credentials from the server config.
Which isn't technically hard, and also usually comes builtin, right in the app/framework of choice.
The Request Manager is already calling the [`Authenticator::set()`](https://github.com/Sammyjo20/Saloon/blob/v1/src/Managers/RequestManager.php#L407).
So, the `Authenticator::set()` methods in Authorization Code Grant could simply check the expiration, and, if expired, request a new Access Token, using the provided Refresh Token.
Client Credential Grant doesn't have this, as these are credentials stored on the server, and no extra work is needed here.

I've edited this message more times than I can count on both hands (and feet), but hopefully I got it right this time, and that it's helpful.
Let me know if I can help with anything. 🙂
I'm very new to collaborating in open source and all, but trying to make my way into it.

**Reply by Sammyjo20** [reply]
Hey @juse-less, thanks for replying that's all super useful information.

I've used Laravel Socialite in the past and it's a great package. I really like your idea of creating an authenticator and running the refreshing logic inside of the "set" method. In my other Oauth 2 integrations, I have been creating a plugin that runs and checks if the authenticator has expired, but your idea feels a lot cleaner.

I could bake in a new feature into Saloon which uses a connector that you create. That way you can configure the default headers, config and the URL of the authorization server, and then you can add a trait called "HasOauthAuthorizationCodeFlow" (or something snapper) which adds all the methods for you, so all you have to do is:

1. Create a connector
2. Add the Oauth 2 trait to the connector
3. Configure all the URLs and everything you need in the connector, for example:

```php
// Inside connector

protected function defineOauthConfig()
{
    return OauthConfig::create('clientId', 'clientSecret', etc, etc);
}
```

5. See methods below

```php
$connector = new TwitterAuthConnector;
$redirectUrl = $connector->getRedirectURL();

$authenticator = $connector->getAccessToken();
$freshAuthenticator = $connector->refreshAccessToken($oldAuthenticator);
```

The full "Oauth" connector could look something like... 

```php
<?php

class TwitterAuthConnector extends SaloonConnector
{
    use HasOauthCodeFlow

    public function defineBaseUrl()
    {
         return 'https://auth.twitter.com';
    }

    public function defineOauthConfig()
    {
       return OauthConfig::create('client-id', 'client-secret');

      // ... Also allow people to set the oauth config outside of the class signature so PHP SDKs can set it 
      // easily
    }

    // ... All other methods provided by trait
}

**Reply by Sammyjo20** [reply]
This is what I am thinking...

![Screen Shot 2022-06-07 at 12 19 05](https://user-images.githubusercontent.com/29132017/172367159-43c70603-7586-48fb-b59c-e9b98ae386f5.png)

**Reply by juse-less** [reply]
I feel like I rushed my idea a bit, so gonna need some time to rethink it a bit.

Now I'm a bit split on whether I still prefer using an `Authenticator`, because of the semantics, or using a `Connector`.
(Although, oAuth is authorization (authz), and not authentication (authn). So technically not semantic to use an `Authenticator`.)

Nonetheless. From the top of my head, I like your way a lot more than my own (which required more of an implementation, like the mentioned The PHP League's oAuth Client library/libraries), than yours, that lets people define the implementation for creating and refreshing tokens, etc.

I'll get back once I've had some time to think about it a bit.

**Reply by Sammyjo20** [reply]
I don't think your idea is bad, I just wanted to work on it. 

I was thinking of passing in the OauthConfig class into the authenticator too so it can be super easily serialised and then stored in the database for example, then retrieved and used again, while maintaining all of the logic you need to refresh tokens. So your idea of using the "set" method could be quite handy here.

**Reply by Sammyjo20** [reply]
So in summary the flow using connectors would be:

1. You create a connector
2. Add the oauth trait to it
3. Define all the default oauth config inside the connector, but it can be overwritten in runtime
4. You run the "createTokens" method 
5. The method will return you an `authenticator` which contains the access token, refresh token, expiry, and a handler for when the token is refreshed
6. You can use that authenticator on any other connector, or the same one. 
7. If a token needs refreshing, then it can refresh and then run the "set" method on the new authenticator

**Reply by Sammyjo20** [reply]
We could even make the refreshing logic entirely optional by only enabling it if a trait is added to the connector/request that the authentication is on, and get people to refresh the tokens manually before making a request, just like the League Oauth 2 client makes you do

from their docs:

<img width="944" alt="Screen Shot 2022-06-07 at 17 25 49" src="https://user-images.githubusercontent.com/29132017/172433404-1a6d3c43-140d-4154-b122-b763afe9979e.png">

**Comment by Sammyjo20** [comment]
What do you guys think of this? I built a simple prototype of what I talked about earlier. I thought I'd use the Spotify API as a good example.

Here is my SpotifyAuthConnector

<img width="698" alt="Screen Shot 2022-06-07 at 21 16 50" src="https://user-images.githubusercontent.com/29132017/172474322-da9deef0-015f-4fc3-92c1-daec99767700.png">

Here is my implementation in a Laravel controller

<img width="804" alt="Screen Shot 2022-06-07 at 21 17 26" src="https://user-images.githubusercontent.com/29132017/172474417-93061901-a1c4-4ea7-8c71-62dd65d69bbc.png">

In this example:
- handleConnect, generating the authorization url
- handleCallback, create the first access token
- handleRefresh, used for refreshing expired tokens

**Comment by Sammyjo20** [comment, marked as answer]
@waapatricio I have released Saloon v1.3 which has some sweet OAuth2 boilerplate that I think you will find useful.

https://docs.saloon.dev/advanced/oauth2-authentication

**Reply by waapatricio** [reply]
Fantastic! Thank you all so much for your collaboration. 🥇

**Reply by waapatricio** [reply]
Also, I would like to share my testimony of how Saloon has helped me in creating SDK packages. Congratulations @Sammyjo20!

**Reply by Sammyjo20** [reply]
Thanks so much @waapatricio, I love the feedback and I love hearing about people using Saloon. Any other feedback? Care to share more about your SDK?

**Reply by juse-less** [reply]
@Sammyjo20 Well done!
I absolutely love how this ended up. 🙂

**Reply by Sammyjo20** [reply]
Thanks @juse-less for your support and your contribution to this, it helped spark the thought for how I wanted it to work and it’s always greatly appreciated. I love that we are all working together to make something awesome.
